### PR TITLE
[AMD] Enable cudagraph for aiter nsa backend and add aiter impl for nsa pr…

### PR DESCRIPTION
**Pull request by wufann**
## Motivation

```
kv_indices = page_table_1[page_table_1 != -1] 
```
Bool mask produced dynamic shape that could not be captured into cuda graph.

Capture NSA decode implemented by aiter backend into cuda graph for better performance.

## Modifications

1. Add a triton kernel get_valid_kv_indices to extract kv indices form transformed paged table.
2. Add a simple implementation with aiter for nsa prefill phase.
## Accuracy Tests

```
# server
python3 -m sglang.launch_server --model /data/models/DeepSeek-V3.2-Exp \
        --tp 8 \
        --port 10086 \
        --trust-remote-code \
        --nsa-prefill-backend aiter \
        --nsa-decode-backend aiter \
        --disable-radix-cache

# client
python3 bench_sglang.py --num-shots 20 --num-questions 1309 --port 10086 --parallel 96

# MI355
Accuracy: 0.958
Invalid: 0.000

# MI300
Accuracy: 0.958
Invalid: 0.000
```
## Benchmarking and Profiling

```
# client
CON="8 16 32 64"
COMBINATIONS=("8192/1024")
for combo in "${COMBINATIONS[@]}"; do
   IFS="/" read -r isl osl <<< "$combo"
   for con in $CON; do
       prompts=$(($con))
       echo "[RUNNING] prompts $prompts isl $isl osl $osl con $con"
       python3 -m sglang.bench_serving \
       --backend sglang \
       --host 127.0.0.1 \
       --port 10086 \
       --dataset-name random \
       --random-range-ratio 1 \
       --num-prompt $((con * 10)) \
       --random-input $isl \
       --random-output $osl \
       --max-concurrency $con
    done
done
```
benchmark on MI355 and compare with tilelang implemention.
| isl  | osl  | concurrency | total throughput(tok/s) |          | median ttft(ms) |          | median tpot(ms) |          |
|------|------|-------------|-------------------------|----------|-----------------|----------|-----------------|----------|
|      |      |             | aiter                   | tilelang | aiter           | tilelang | aiter           | tilelang |
| 8192 | 1024 | 8           | 1784.17                 | 1087.3   | 3443.55         | 4994.87  | 37.04           | 61.49    |
| 8192 | 1024 | 16          | 2878.6                  | 1841.45  | 5374.7          | 8365.29  | 44.79           | 70.05    |
| 8192 | 1024 | 32          | 4085.2                  | 2762.07  | 9273.7          | 15119.79 | 61.46           | 89.62    |
| 8192 | 1024 | 64          | 5283.04                 | 3759.46  | 17054.57        | 28497.06 | 92.43           | 125.47   |

profiling decode phase:
aiter
Pasted image 20260210130258
tilelang
Pasted image 20260210130426

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

cc: @hubertlu-tw @kkHuang-amd

### CODE REVIEW

**Review by hubertlu-tw — APPROVED** [review]
LGTM

File: python/sglang/srt/layers/attention/nsa_backend.py [hunk 1]
```diff
@@ -1750,6 +1768,54 @@ def _forward_aiter(
...
+        kv_indptr = torch.zeros(num_tokens + 1, dtype=torch.int32, device=self.device)
+        kv_indptr[1:] = torch.cumsum(non_minus1_counts, dim=0)
+
+        # Compact valid global indices
+        total_valid = kv_indptr[-1].item()
+        kv_indices = torch.zeros(total_valid, dtype=torch.int32, device=self.device)
+
+        # Use get_valid_kv_indices kernel to extract valid indices
+        if num_tokens > 0 and total_valid > 0:
+            get_valid_kv_indices(page_table_1, kv_indptr, kv_indices, num_tokens)
```
Comment (kkHuang-amd) [inline python/sglang/srt/layers/attention/nsa_backend.py]: Line #1798 will have potential issue.

The kv_indices size is total_valid, but get_valid_kv_indices needs the kv_indices size is (bs * topk).

total_valid = (bs * topk) only when the kv length is >= topk, if kv lenght < topk, total_valid is not equal to (bs * topk)

Comment (wufann) [reply python/sglang/srt/layers/attention/nsa_backend.py]: Fixed. Pre allocate kv_indices with size (bs * topk) for safety.

**Review by HaiShaw — APPROVED** [review]
